myia/api.py

```diff
@@ -503,3 +511,89 @@ def step(self, graph):
 scalar_debug_compile = scalar_debug_pipeline \
     .select('parse', 'resolve', 'export') \
     .make_transformer('input', 'output')
+
+
+#################
+# Top-level API #
+#################
+
+
+class MyiaFunction:
+    """Represents a function compiled by Myia.
+
+    MyiaFunction will compile the original function for every combination of
+    argument types and shapes it is given (as well as their values,
+    optionally).
+
+    Attributes:
+        fn: The root function to compile.
+        specialize_values: Set of arguments for which we should specialize the
+            function based on their values (list of argument names).
+
+    """
+
+    def __init__(self, fn, specialize_values=[]):
+        """Initialize a MyiaFunction."""
+        self.fn = fn
+        self.specialize_values = set(specialize_values)
+        self._cache = {}
+
+    def specialize(self, args):
+        """Specialize on the types of the given arguments.
+
+        Returns a Pipeline. If the argument types were seen before, returns a
+        cached version.
+        """
+        pip = standard_debug_pipeline.make()
+        inf = pip.steps.infer
+        argspec = tuple({'value': arg} for arg in args)
+        inf.fill_in(argspec)
+        argnames = inspect.getargspec(self.fn).args
+        for arg, name in zip(argspec, argnames):
```

Comment:
It works fine. The loop does not change the fact that `len(argspec) == len(args)`, and the inferrer will verify that `argspec` has the proper length. I think it would be redundant to add a check here.

Reply:
Yes, on closer inspection, you're right.  Can you at least add a test for too many/too few arguments.

Reply:
Done.